README.md
# ACL
ACL library provided for example/use with [Rumbleship GQL Framework](https://github.com/rumbleship/rfi-gql-framework/)

## The Four Questions
Authentication & Authorization have four critical questions:
1. Authenticate -> Do I exist? -> Am I someone in the Rumbleship system?
    > failure: 401; "bad password", "your credentials have been revoked"

2. Scope / Resolver -> Can I ask this kind of question? -> Do I have access to ask for data from this particular table in the database?
    > failure: 403; "you're not allowed to ask about BankAccounts"

3. Instance Filtering: Sequelize Service -> Can I ask this specific question? -> Loads instances from database based on passed in filter
    > expected failure: 200 with no empty lists; "you can ask about bank accounts, but there are no bank accounts you can see".

    > unexpected failure: 500. true error.

4. Attribute Filtering: Relay -> Can I ask for these specific fields/details on an instance? -> whitelists specific attributes on the instance
    > expected failure: 200 with data, where unallowed fields are `null`; "you can ask about bank accounts, and here are five of them you can see, but you're not allowed to view `account_number` on any of them, so we've nulled it.

    > unexpected failure: 500. true error.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. Starting with v0.1.0.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [UNRELEASED]

### Added
  * MIT License
### Removed
  * tslint
### Changed
  * eslint
  * upgrad to typescript 4
### Fixed
### Deprecated
### Security


## [2.2.0] -- 2020-08-28

### Added
  * support for `on_behalf_of` claim

## [2.1.0] -- 2020-08-19

### Added
  * Scopes.MIGRATION

## [2.0.1] -- 2020-07-30

### Changed
  * `createSysAdminAuthHeader()` --> `createServiceUserAuthHeader()`

## [2.0.0] -- 2020-07-30

### Added
  * Guarantee the presence of `claims.user`: https://www.pivotaltracker.com/story/show/174020936
    * static `Authorizer.initialize()` to allow removing boilerplate passing of `SECRET`
    * peer dependency on `@rumbleship/config` for type interfaces
### Removed
  * `createAuthHeader()` and `createRefreshHeader` in favor of static methods: `Authorizer.createAuthHeader()` and `Authorizer.createRefreshHeader()`
  * deprecated methods; getRoles, getClient
### Changed
  * RefreshClaims has `user` instead of `owner` for consistency with `AccessClaims`
### Fixed
  * launch.json boilerplate for new debugger

## [1.4.0] -- 2020-07-15

### Changed
  * Roll logic for `authorizerTreatAs` to return either with inflected ids, or without -- defaults to with

## [1.3.0] -- 2020-07-10

### Changed
  * @AuthorizerTreatAs accepts either a single resource or a list of resources

## [1.2.1] -- 2020-06-05

### Added
  * `NEW_JWT_THOUGHTS.md` to collect thoughts on how we may change the jwt to have a deeper, more granular, more data-driven permission structure.
  * `Scopes.API_KEY` to group together the permissions that an `ApiKey` has, but a `User` does not have.

## [1.2.0] -- 2020-04-20

### Removed
  * peer dep on `@rumbleship/spyglass`
### Changed
  * package.json engine supports node 12
  * circle builds on node12
  * Upgrade to `@rumbleship/oid#4.0.2`

## [1.1.1] -- 2020-03-31

### Removed
  * Approve for publishing build step on merge to master
  * `xxhash` dependency
### Changed
  * Updated tslint.json, tsconfig.json, plugins, etc to support nullish coalescing.
  * Upgrade to `@rumbleship/acl#4.0.1`

## [1.1.0] -- 2020-03-05

### Changed
  * Throw `InvalidJWTError` instead of plain `Error` when constructing an Authorizer with invalid accessToken: https://www.pivotaltracker.com/story/show/171657687

## [1.0.3] -- 2020-01-21

### Changed
  * o11y fix: authorizer.getUser(), .getRoles() will still return after authenticating an accesstoken -- even if the token has expired

## [1.0.2] -- 2019-12-24

### Fixed
  * authorizer.authenticate() no longer throws if passed an access token that has a falsy role object

## [1.0.1] -- 2019-12-23

### Added
  * Scopes.USER

## [1.0.0] -- 2019-12-15

### Added
  * `AuthorizerTreatAs` property decorator and `getAuthorizerTreatAs()` pair, allowing explicit mapping of a property to the resource it abstracts, e.g.:
    * `owner_id` to `Resource.User`
    * `buyer_division_id`, `supplier_division_id` to `Resource.Division`
    * TODO: make this a parameterDecorator as well
### Removed
  * `AuthorizedResource` and `AuthorizedAttribute` decorators
### Changed
  * `authorizer.can()` accepts explicit attr/resource overrides as a map; defaults to collating information from decorators

## [0.0.2] -- 2019-08-14

### Added
  * DivisionAdmin scope
  * `.prettierignore` file
  * Skeleton for generating and consuming `refreshTokens`
  * Decorators for injecting the `attribute` and `resource` into a subclass's method definition
  * keys and values in `Actions` enum relating to `OrderEvent`
### Removed
  * `moment` is no longer a peer dependency, just a `devDependency`
### Changed
  * `Actions.PURCHASE`, `Actions.SHIP` and `Actions.RETURN` values are now uppercase
  * `can` accepts either an attribute or a list of attributes

## [0.0.1] -- 2019-07-17

### Added

- Initial Release; parity with what is currently executed in Alpha

### Removed

### Changed

### Fixed

### Deprecated

### Security


lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrantTypes = exports.Scopes = exports.Resource = exports.Actions = exports.Roles = void 0;
var Roles;
(function (Roles) {
    Roles["ADMIN"] = "admin";
    Roles["USER"] = "user";
    Roles["PENDING"] = "pending";
})(Roles = exports.Roles || (exports.Roles = {}));
var Actions;
(function (Actions) {
    Actions["CREATE"] = "create";
    Actions["READ"] = "read";
    Actions["UPDATE"] = "update";
    Actions["DELETE"] = "delete";
    Actions["QUERY"] = "query";
    // are Approve and Verify the same action, e.g. a limited update?
    Actions["APPROVE"] = "approve";
    Actions["VERIFY"] = "verify";
    Actions["REQUEST"] = "request";
    // these values are uppercased to be compatible with keys to make it easier to access values without having to change the case --2019-08-06 @preeV42500
    Actions["PURCHASE"] = "PURCHASE";
    Actions["SHIP"] = "SHIP";
    Actions["RETURN"] = "RETURN";
})(Actions = exports.Actions || (exports.Actions = {}));
var Resource;
(function (Resource) {
    Resource["Division"] = "Division";
    Resource["User"] = "User";
    Resource["Order"] = "Order";
})(Resource = exports.Resource || (exports.Resource = {}));
var Scopes;
(function (Scopes) {
    Scopes["MIGRATION"] = "migration";
    Scopes["USER"] = "user";
    Scopes["API_KEY"] = "api_key";
    Scopes["SYSADMIN"] = "system:*";
    Scopes["BANKINGADMIN"] = "banking:*";
    Scopes["ORDERADMIN"] = "orders:*";
    Scopes["DIVISIONADMIN"] = "divisions:*";
})(Scopes = exports.Scopes || (exports.Scopes = {}));
var GrantTypes;
(function (GrantTypes) {
    GrantTypes["REFRESH"] = "refresh";
    GrantTypes["ACCESS"] = "access";
})(GrantTypes = exports.GrantTypes || (exports.GrantTypes = {}));
//# sourceMappingURL=types.js.map